File: 3_SparseMatrix/Sequential/SparseMatrix.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdbool.h>

int main() {
    srand(time(NULL));

    int n, m;
    int zeros = 0;

    printf("--------- SparseMatrix&Sequential --------\n");

    printf("Ingrese la cantidad de filas de la matriz: ");
    scanf("%d", &n);

    printf("Ingrese la cantidad de columnas de la matriz: ");
    scanf("%d", &m);

    int matrix[n][m];

    printf("\nLa matriz generada al azar es:\n");
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            matrix[i][j] = rand() % 2;
            printf("%d ", matrix[i][j]);
        }
        printf("\n");
    }

    clock_t start = clock();

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (matrix[i][j] == 0)
                zeros++;
        }
    }

    clock_t end = clock();

    if (zeros > (n * m) / 2)
        printf("\nLa matriz es dispersa (sparse)\n");
    else
        printf("\nLa matriz no es dispersa\n");

    printf("Cantidad de ceros: %d de %d elementos\n", zeros, n * m);
    printf("Tiempo total de ejecución: %.6f segundos\n",
           (double)(end - start) / CLOCKS_PER_SEC);

    printf("---------- Program End ----------\n");
    return 0;
}

File: Python/EntradaAleatoria/3_SparseMatrix/Sequential/sparse_matrix.py
#!/usr/bin/env python3
"""
SparseMatrix sequential - EntradaAleatoria (Python, numpy)
Reads n m from params.txt if present, otherwise uses defaults. Generates a random 0/1 matrix,
prints it, counts zeros and reports whether it's sparse (>50% zeros). Uses CPU timing (time.process_time()).
"""
import sys
import time
try:
    import numpy as np
except ImportError:
    print('numpy is required for this script. Install with: pip install numpy')
    sys.exit(1)

DEFAULT_N = 10
DEFAULT_M = 10

def read_dims(path='params.txt'):
    try:
        with open(path,'r') as f:
            toks = f.read().strip().split()
            if len(toks) >= 2:
                return int(toks[0]), int(toks[1])
    except Exception:
        pass
    return None, None

if __name__ == '__main__':
    n,m = read_dims()
    if n is None or m is None:
        n,m = DEFAULT_N, DEFAULT_M
    print('--------- SparseMatrix&Sequential (EntradaAleatoria - Python) --------')
    mat = np.random.randint(0,2,size=(n,m), dtype='i')
    print('\nLa matriz generada es:')
    for i in range(n):
        print(' '.join(str(int(x)) for x in mat[i]))

    start = time.process_time()
    zeros = int((mat == 0).sum())
    end = time.process_time()

    total = n*m
    print(f"\nCantidad total de ceros: {zeros} de {total} elementos")
    if zeros > total/2:
        print('La matriz es dispersa (sparse)')
    else:
        print('La matriz no es dispersa')
    print(f"Tiempo total de ejecución (CPU): {end - start:.6f} segundos")
    print('---------- Program End ----------')
